server/app/controllers/applicant/ApplicantProgramBlocksController.java

```diff
@@ -461,92 +461,137 @@ private void removeAddressJsonFromSession(Request request) {
     request.session().removing(ADDRESS_JSON_SESSION_KEY);
   }
 
-  /** Navigates to the previous page of the application. */
+  /**
+   * Navigates to the previous page of the application for a specific applicant.
+   *
+   * @param request the HTTP request containing context and session information
+   * @param applicantId the unique identifier of the applicant whose application is being navigated
+   * @param programParam the program identifier, which can be either a program slug or program ID,
+   *     depending on feature configuration
+   * @param previousBlockIndex the index of the previous block to navigate to within the program's
+   *     block sequence
+   * @param inReview indicates whether the applicant is currently in review mode (true) or edit mode
+   *     (false), which affects navigation behavior and view rendering
+   * @param isFromUrlCall indicates whether this method was invoked directly from a URL call, used
+   *     to determine redirect behavior when program slug URLs are enabled
+   * @return a CompletionStage that resolves to a Result containing the rendered previous block
+   *     page, or error responses (unauthorized, not found, or redirect) if authorization fails or
+   *     resources are not found
+   */
   @Secure
   public CompletionStage<Result> previousWithApplicantId(
-      Request request, long applicantId, long programId, int previousBlockIndex, boolean inReview) {
-    CompletionStage<ApplicantPersonalInfo> applicantStage =
-        this.applicantService.getPersonalInfo(applicantId);
-
-    CompletableFuture<Void> applicantAuthCompletableFuture =
-        applicantStage
-            .thenComposeAsync(
-                v -> checkApplicantAuthorization(request, applicantId),
-                classLoaderExecutionContext.current())
-            .toCompletableFuture();
+      Request request,
+      long applicantId,
+      String programParam,
+      int previousBlockIndex,
+      boolean inReview,
+      boolean isFromUrlCall) {
+    // Redirect home when the program param is the program id (numeric) but it should be the program
+    // slug because the program slug URL is enabled and it comes from the URL call
+    boolean programSlugUrlEnabled = settingsManifest.getProgramSlugUrlsEnabled(request);
+    if (programSlugUrlEnabled && isFromUrlCall && StringUtils.isNumeric(programParam)) {
+      metricCounters
+          .getUrlWithProgramIdCall()
+          .labels(
+              "/applicants/:applicantId/programs/:programParam/blocks/:previousBlockIndex/previous/:inReview",
+              programParam)
+          .inc();
+      return CompletableFuture.completedFuture(redirectToHome());
+    }
 
-    CiviFormProfile profile = profileUtils.currentUserProfile(request);
+    return programSlugHandler
+        .resolveProgramParam(programParam, applicantId, isFromUrlCall, programSlugUrlEnabled)
+        .thenCompose(
+            programId -> {
```

Comment:
All this is the same as before, just moved after resolving the program param (with a couple clean ups for early exit)